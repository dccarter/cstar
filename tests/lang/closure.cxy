
func main(args: [string]) {
    // simple closure
    const one = () => 1;
    one();

    // iic
    ((name: string)  => {
        return f"Hello ${name}"
    })("User");

    // capture
    var x = 100;
    var b = "Hello";
    (() => f"${b} -> ${x}")()
}