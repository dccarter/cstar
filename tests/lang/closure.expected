{
 "tag": "Program",
 "decls": [{
   "tag": "FuncDecl",
   "flags": "(TopLevelDecl)",
   "overload": 51,
   "index": 0,
   "name": "main",
   "params": [{
     "tag": "FuncParam",
     "name": "args",
     "index": 0,
     "paramType": {
      "tag": "ArrayType",
      "flags": "(TypeAst)",
      "element": {
       "tag": "StringType",
       "flags": "(TypeAst)"
      }
     }
    }],
   "body": {
    "tag": "BlockStmt",
    "stmts": [{
      "tag": "VarDecl",
      "flags": "(Const)",
      "names": [{
        "tag": "Identifier",
        "value": "one"
       }],
      "init": {
       "tag": "ClosureExpr",
       "body": {
        "tag": "IntegerLit",
        "value": 1
       }
      }
     }, {
      "tag": "ExprStmt",
      "expr": {
       "tag": "CallExpr",
       "callee": {
        "tag": "Path",
        "elements": [{
          "tag": "PathElem",
          "name": "one",
          "index": 0
         }],
        "isType": false
       },
       "overload": 0
      }
     }, {
      "tag": "ExprStmt",
      "expr": {
       "tag": "CallExpr",
       "callee": {
        "tag": "GroupExpr",
        "expr": {
         "tag": "ClosureExpr",
         "params": [{
           "tag": "FuncParam",
           "name": "name",
           "index": 0,
           "paramType": {
            "tag": "StringType",
            "flags": "(TypeAst)"
           }
          }],
         "body": {
          "tag": "BlockStmt",
          "stmts": [{
            "tag": "ReturnStmt",
            "expr": {
             "tag": "StringExpr",
             "parts": [{
               "tag": "StringLit",
               "value": "Hello "
              }, {
               "tag": "Path",
               "elements": [{
                 "tag": "PathElem",
                 "name": "name",
                 "index": 0
                }],
               "isType": false
              }, {
               "tag": "StringLit",
               "value": ""
              }]
            }
           }]
         }
        }
       },
       "args": [{
         "tag": "StringLit",
         "value": "User"
        }],
       "overload": 0
      }
     }, {
      "tag": "VarDecl",
      "names": [{
        "tag": "Identifier",
        "value": "x"
       }],
      "init": {
       "tag": "IntegerLit",
       "value": 100
      }
     }, {
      "tag": "VarDecl",
      "names": [{
        "tag": "Identifier",
        "value": "b"
       }],
      "init": {
       "tag": "StringLit",
       "value": "Hello"
      }
     }, {
      "tag": "ExprStmt",
      "expr": {
       "tag": "CallExpr",
       "callee": {
        "tag": "GroupExpr",
        "expr": {
         "tag": "ClosureExpr",
         "body": {
          "tag": "StringExpr",
          "parts": [{
            "tag": "StringLit",
            "value": ""
           }, {
            "tag": "Path",
            "elements": [{
              "tag": "PathElem",
              "name": "b",
              "index": 0
             }],
            "isType": false
           }, {
            "tag": "StringLit",
            "value": " -> "
           }, {
            "tag": "Path",
            "elements": [{
              "tag": "PathElem",
              "name": "x",
              "index": 0
             }],
            "isType": false
           }, {
            "tag": "StringLit",
            "value": ""
           }]
         }
        }
       },
       "overload": 0
      }
     }]
   }
  }]
}
